v1/plugins/logs/eventBuffer.go

```diff
@@ -0,0 +1,259 @@
+package logs
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/open-policy-agent/opa/v1/plugins/rest"
+)
+
+// eventBuffer stores and uploads a gzip compressed JSON array of EventV1 entries
+type eventBuffer struct {
+	buffer               chan []byte    // buffer stores JSON encoded EventV1 data
+	mtx                  sync.RWMutex   // mtx controls access to the buffer
+	payload              *payloadBuffer // payload contains the compressed JSON array to be uploaded
+	client               rest.Client    // client is used to upload the data to the configured service
+	uploadPath           string         // uploadPath is the configured HTTP resource path for upload
+	uploadSizeLimitBytes int64          // uploadSizeLimitBytes will enforce a maximum payload size to be uploaded
+}
+
+func newEventBuffer(bufferSizeLimitEvents int64, client rest.Client, uploadPath string, uploadSizeLimitBytes int64) *eventBuffer {
+	return &eventBuffer{
+		buffer:               make(chan []byte, bufferSizeLimitEvents),
+		payload:              newPayloadBuffer(),
+		client:               client,
+		uploadPath:           uploadPath,
+		uploadSizeLimitBytes: uploadSizeLimitBytes,
+	}
+}
+
+// Reconfigure updates the user configurable values
+func (b *eventBuffer) Reconfigure(bufferSizeLimitEvents int64, client rest.Client, uploadPath string, uploadSizeLimitBytes int64) (int, []error) {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
+	b.client = client
+	b.uploadPath = uploadPath
+	oldUploadSizeLimitBytes := b.uploadSizeLimitBytes
+	b.uploadSizeLimitBytes = uploadSizeLimitBytes
+
+	if int64(cap(b.buffer)) == bufferSizeLimitEvents && uploadSizeLimitBytes == oldUploadSizeLimitBytes {
+		return 0, nil
+	}
+
+	close(b.buffer)
+	newBuffer := make(chan []byte, bufferSizeLimitEvents)
+
+	var errs []error
+	var dropped int
+	for event := range b.buffer {
+		if int64(len(event)) < b.uploadSizeLimitBytes {
+			if push(newBuffer, event) {
+				dropped++
+			}
+			continue
+		}
+
+		var decoded EventV1
+		if err := json.Unmarshal(event, &decoded); err != nil {
+			dropped++
+			errs = append(errs, err)
+			continue
+		}
+		if err := b.dropNDCache(&decoded, event); err != nil {
+			if !errors.Is(err, droppedNDCache{}) {
+				dropped++
+				errs = append(errs, err)
+				continue
+			}
+			event, err = json.Marshal(decoded)
+			if err != nil {
+				dropped++
+				errs = append(errs, err)
+				continue
+			}
+		}
+		if push(newBuffer, event) {
+			dropped++
+		}
+	}
+	b.buffer = newBuffer
+
+	return dropped, errs
+}
+
+// Push attempts to add a new event to the buffer, returning true if an event was dropped.
+func (b *eventBuffer) Push(event EventV1) (bool, error) {
+	encoded, err := json.Marshal(event)
```

Comment:
So we're marshalling the event before putting it on the buffer? Per-request performance would likely increase if this was done on the consumer-side.

Reply:
Especially considering the worst-case scenario, where `dropNDCache` marshalls a second time.

Reply:
True but the downside would be that the ND cache will then have to be dropped at the consumer side. Which could mean an event that is too big to upload will take space in the buffer for it only to be dropped on upload. Probably better to get rid of an event as soon as possible?

Reply:
Agreed, updated the PR to move the serialization to the consumer side 👍